kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaAuthnHandler.java

```diff
@@ -0,0 +1,388 @@
+
+        private SaslMechanism(String saslName, ScramMechanism scramMechanism,
+                              String... negotiableProperties) {
+            this.name = saslName;
+            this.scramMechanism = scramMechanism;
+            this.negotiableProperties = negotiableProperties;
+        }
+
+        public String mechanismName() {
+            return name;
+        }
+
+        static SaslMechanism fromMechanismName(String mechanismName) {
+            switch (mechanismName) {
+                case "PLAIN":
+                    return PLAIN;
+                case "SCRAM-SHA-256":
+                    return SCRAM_SHA_256;
+                case "SCRAM-SHA-512":
+                    return SCRAM_SHA_512;
+            }
+            throw new UnsupportedSaslMechanismException(mechanismName);
+        }
+
+        public ScramMechanism scramMechanism() {
+            return scramMechanism;
+        }
+
+        public String[] negotiableProperties() {
+            return negotiableProperties;
+        }
+    }
+
+    // TODO need some way to support SaslAuthenticate v0, which doesn't use Kafka protcol framing at all
+
+    private final List<String> enabledMechanisms;
+
+    @VisibleForTesting
+    SaslServer saslServer;
+
+    private final Map<String, AuthenticateCallbackHandler> mechanismHandlers;
+
+    private State lastSeen;
+
+    public KafkaAuthnHandler(Map<KafkaAuthnHandler.SaslMechanism, AuthenticateCallbackHandler> mechanismHandlers) {
+        this(State.START, mechanismHandlers);
+    }
+
+    @VisibleForTesting
+    KafkaAuthnHandler(State init,
+                      Map<KafkaAuthnHandler.SaslMechanism, AuthenticateCallbackHandler> mechanismHandlers) {
+        this.lastSeen = init;
+        this.mechanismHandlers = mechanismHandlers.entrySet().stream().collect(Collectors.toMap(
+                e -> e.getKey().mechanismName(), Map.Entry::getValue));
+        this.enabledMechanisms = List.copyOf(this.mechanismHandlers.keySet());
+    }
+
+    IllegalStateException illegalTransition(State next) {
+        return new IllegalStateException("Illegal state transition from " + lastSeen + " to " + next);
+    }
+
+    void validateTransition(State next) {
+        State previous = lastSeen;
+        switch (next) {
+            case API_VERSIONS:
+                if (previous != State.START) {
+                    throw illegalTransition(next);
+                }
+                break;
+            case SASL_HANDSHAKE_v0:
+            case SASL_HANDSHAKE_v1_PLUS:
+                if (previous != State.START
+                        && previous != State.API_VERSIONS) {
+                    throw illegalTransition(next);
+                }
+                break;
+            case UNFRAMED_SASL_AUTHENTICATE:
+                if (previous != State.START
+                        && previous != State.SASL_HANDSHAKE_v0
+                        && previous != State.UNFRAMED_SASL_AUTHENTICATE) {
```

Comment:
why allow the state transition to itself ?  UNFRAMED_SASL_AUTHENTICATE -> UNFRAMED_SASL_AUTHENTICATE

Reply:
SCRAM-SHA requires two request/response interactions. `SaslAuthenticate` appeared in Kafka 1.0 (via KIP-152), but SASL SCRAM support existed from 0.10.2.0 (via KIP-84), therefore there exist (or _existed_) clients which used SCRAM-SHA mechanisms without encapsulation in `SaslAuthenticate` frames. Of course it's entirely legitimate to wonder how many such clients are still in use.

Reply:
thanks for the explanation.